Switch to Agency

lchanmann opened the pull request:
~Stub thread pool would not execute context agent upon submission. To actually run the submitted agents, the client needs to call `invokeAll()` manually.~

## Changes

- Replace ThreadPool with Agency
- Rename FixedThreadPool to ThreadPoolAgency
- Add AgentAgency (it was StubThreadPool before)
- Update "threadPool" service setting

lchanmann committed:
Cast to ThreadPool interfact instead of FixedThreadPool class

lchanmann committed:
Implement StubThreadPool

jlhughes commented:
Who (or will) calls invokeAll?

lchanmann commented:
> Who (or will) calls invokeAll?

I plan to use it for IdentityMind integration test. It allows us to invoke the real (async) actions before performing the assertions. For example:

```java
((IdentityMindService) getIdentityMindService()).setDefaultProfile("ACCEPT");
Transaction tx = newTransaction(1000);
tx = (Transaction) ((DAO) getTransactionDAO()).put(tx);

test(findTxApprovalRequest(tx) == null,
  "No approval request is created yet before the IdentityMind transaction rule is run");

// Invoke IdentityMind validator
((StubThreadPool) getThreadPool()).invokeAll();

// Assertions
ApprovalRequest found = findTxApprovalRequest(tx);
test(found != null && found.getApprover() == getIdentityMindUserId() && found.getStatus() == ApprovalStatus.APPROVED,
  "An approval request with status=APPROVED is created when the IdentityMind ACCEPT-ed the transaction");

tx = (Transaction) ((DAO) getTransactionDAO()).find(tx);
test(tx.getStatus() == TransactionStatus.COMPLETED, "Compliance transaction should be COMPLETED");
```

kgrgreer commented on src/foam/nanos/pool/ThreadPool.java:
```diff
@@ -10,5 +10,5 @@
 import foam.core.X;
 
 public interface ThreadPool {
```
Isn't this just the same as Agency?

kgrgreer commented:
Isn't threadpool the same interface as Agency and FixedThreadPool should just be an Agency implementation?

lchanmann committed:
Merge branch master into AddStubThreadPool

lchanmann committed:
Cleanup Agency

- Remove Agency.java
- Create Agency interface model with only one `submit(x, agent, description)` method
- Update other classes that use Agency

lchanmann committed:
Convert StubThreadPool to AgentAgency

lchanmann committed:
Switch from ThreadPool to Agency

- Remove ThreadPool interface
- Convert FixedThreadPool to ThreadPoolAgency
- Update services file and other classes that have references to ThreadPool

lchanmann commented:
Added your suggestions. @kgrgreer

lchanmann committed:
Merge remote-tracking branch 'upstream/master' into AddStubThreadPool

kgrgreer commented on src/foam/nanos/pool/AgentAgency.java:
```diff
@@ -0,0 +1,24 @@
+
+  public void execute() {
+    while ( ! agents.isEmpty() ) {
+      agents.poll().run();
```
Why poll?

lchanmann replied:
Fixed.

lchanmann committed:
Change from poll() to remove()
